src/orion/core/io/database/ephemeraldb.py

```diff
@@ -0,0 +1,440 @@
+        """Drop the collection, removing all documents and indexes."""
+        self._documents = []
+        self._indexes = dict()
+
+
+class EphemeralDocument(object):
+    """Non permanent document
+
+    This document is meant for debugging purposes within the EphemeralDB.
+
+    .. seealso:: :class:`orion.core.io.database.ephemeraldb.EphemeralDB` for database object.
+
+    """
+
+    operators = {
+        "$in": (lambda a, b: a in b),
+        "$gte": (lambda a, b: a >= b),
+        "$gt": (lambda a, b: a > b)
+    }
+
+    def __init__(self, data):
+        """Initialise the document with a flattened version of the data"""
+        self._data = _flatten(data)
+
+    def match(self, query=None):
+        """Test if the document corresponds to a given query"""
+        if query is None or query == {}:
+            return True
+
+        query = _flatten(query)
+        for key, value in query.items():
+            if not self.match_key(key, value):
+                return False
+
+        return True
+
+    def match_key(self, key, value):
+        """Test if a data corresponding to the given key is in agreement with the given
+        value based on the operator defined within the key.
+
+        Default operator is equal when no operator is defined.
+        Other operators could be $in, $gte, $gt. They are defined
+        in the last section of the key. For example: `abc.def.$in` or `abc.def.$gte`.
+        """
+        if key.split(".")[-1] in self.operators:
+            operator = key.split(".")[-1]
+            key = ".".join(key.split(".")[:-1])
+
+            return key in self and self.operators[operator](self[key], value)
+
+        return key in self and self[key] == value
+
+    def _validate_keys(self, keys):
+        """Verify that all keys are 0 or 1 (with exception of _id) and convert them.
+
+        For simplicity, when keys are 0, the inverse set of keys for 1s is computed.
+
+        .. note ::
+
+            _id is set to 1 if not specified. Only _id may be set to 0 if other keys are set to 1.
+        """
+        keys_without_id = [key for key in keys if key != '_id']
+        n_keys = sum(keys[key] for key in keys_without_id)
+        if n_keys != 0 and n_keys != len(keys_without_id):
+            raise ValueError(
+                'Cannot mix selection with 1 and 0s except for _id: {}'.format(keys))
+
+        if n_keys == 0:
+            new_keys = dict((key, 1) for key in self._data.keys() if key not in keys)
+            new_keys['_id'] = keys.get('_id', 1)
+            keys = new_keys
+
+        return keys
+
+    def select(self, keys):
+        """Only select or only drop the specified keys
+
+        For a pair (key, value) in the dictionnary, value=0 means the key will not be included
+        while value=1 means it will.
+
+        All specified keys should be 0 or 1. They cannot have different values with the exception
+        of _id which can be specified to 0 while the others are at 1.
```

Comment:
Is this due to simplicity in the mock's implementation?

Reply:
As for `upsert` this is to mimic how it works in mongodb.

Ref: https://docs.mongodb.com/manual/reference/method/db.collection.find/#projection